### PR TITLE
Dev pm

### PR DESCRIPTION
**Pull request by p-mayor**
Relevant issue: https://github.com/GrafeasGroup/tor_ocr/issues/30

## Description:

Acceptance Criteria:
- Blossom's record of a post is marked as OCR attempted if call to OCR service succeeds (regardless of if it found content)
- Blossom's record of a post is not marked as OCR attempted if a call to OCR service fails
Updated flow:
- take the ID that we get from Redis when u/ToR puts in a new post to create an OCR attempt from
- ask Blossom for data at the submission endpoint for this ID
- verify that has_ocr_transcription is False
- create and post Transcription object to Blossom for the above submission ID

## Screenshots:
example transcription record:
Screen Shot 2020-01-27 at 10 51 11 PM

## Testing Instructions:

- start a local instance of blossom api
  - make sure there is a submission at submission/1 as this is the hard-coded submission for testing
- start a redis server
- start the tor_ocr bot with `tor-apprentice
- open redis cli
  - add reddit ids for post to transcribe and place to post transcription
  - set [post to transcribe] [place to post transcription]
  - e.g. `set t3_eqhr0b t3_eqlhj4`
  - now push these ids to the ocr_ids key
  - e.g. `rpush ocr_ids t3_eqhr0b`

tor_ocr bot should see the new post and get an ocr transcription for it. It will then check the api for the submission and then check if the submission's 'has_ocr_transcription' field is false. If it is false, a new transcription record will be created.

## Checklist:

- [x] Code Quality
- [ ] Pep-8
- [ ] Tests (if applicable)
- [x] Success Criteria Met
- [ ] Inline Documentation
- [ ] Wiki Documentation (if applicable)

### CODE REVIEW

File: tor_ocr/main.py [hunk 1]
```diff
@@ -36,6 +38,17 @@
...
+dotenv.load_dotenv()
+# TODO: set the client credentials to environment variables
+b_api = BlossomAPI(
+    email=os.environ.get('TOR_OCR_EMAIL'), 
+    password=os.environ.get('TOR_OCR_PASSWORD'), 
+    api_key='TOR_OCR_BLOSSOM_API_KEY', 
+    api_base_url=os.environ.get('TOR_OCR_BLOSSOM_API_BASE_URL')
+)
+
+api_base_url: str=os.environ.get('TOR_OCR_BLOSSOM_API_BASE_URL'),
```
Comment (itsthejoker) [inline tor_ocr/main.py]: Do these lines need to be here or were they references?

File: tor_ocr/main.py [hunk 2]
```diff
@@ -36,6 +38,17 @@
 
     u_tor_post_id.reply(ocr_magic)
 """
+dotenv.load_dotenv()
+# TODO: set the client credentials to environment variables
```
Comment (itsthejoker) [inline tor_ocr/main.py]: Looks like this one's complete!

File: tor_ocr/main.py [hunk 3]
```diff
@@ -245,11 +258,14 @@ def run(config):
         _(base_comment.format(result['process_time_in_ms'] / 1000))
     )
 
-    for chunk in chunks(result['text'], 9000):
+    for counter, chunk in enumerate(chunks(result['text'], 9000)):
+        # grab first comment for transcription url
+        if counter==0:
```
Comment (itsthejoker) [inline tor_ocr/main.py]: Will putting it here grab the introduction comment or the first comment of the generated transcription?

File: tor_ocr/main.py [hunk 4]
```diff
@@ -268,12 +303,25 @@ def noop(cfg):
     logging.info('Loop!')
 
 
+def get_volunteer_id(reddit_username):
+    #remove first 2 characters (eg: u/transcribot => transcribot)
```
Comment (itsthejoker) [inline tor_ocr/main.py]: Great comment!

File: tor_ocr/main.py (hunk 1 above)
Comment (p-mayor) [reply tor_ocr/main.py]: reference, should be removed now.

File: tor_ocr/main.py (hunk 2 above)
Comment (p-mayor) [reply tor_ocr/main.py]: yep, removed todo comment

File: tor_ocr/main.py (hunk 3 above)
Comment (p-mayor) [reply tor_ocr/main.py]: good catch, its getting the intro comment there. I moved it back and tested to make sure its getting the first transcription comment.

File: tor_ocr/core/blossom.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            api_key: str=None,
+            api_base_url: str="http://api.grafeas.localhost:8000",
+            login_url: str="http://grafeas.localhost:8000/login/"
+    ):
+        if not email:
+            raise ConfigurationError("Need an email address!")
+        if not password:
+            raise ConfigurationError("Need a password!")
+        if not api_key:
+            raise ConfigurationError("Need a password!")
```
Comment (itsthejoker) [inline tor_ocr/core/blossom.py]: This is definitely my fault, but this error message should read "Need an API key!"

File: tor_ocr/main.py [hunk 6]
```diff
@@ -36,6 +38,14 @@
...
     u_tor_post_id.reply(ocr_magic)
 """
+dotenv.load_dotenv()
+
+b_api = BlossomAPI(
+    email=os.environ.get('TOR_OCR_EMAIL'), 
+    password=os.environ.get('TOR_OCR_PASSWORD'), 
+    api_key='TOR_OCR_BLOSSOM_API_KEY', 
+    api_base_url=os.environ.get('TOR_OCR_BLOSSOM_API_BASE_URL')
+)
```
Comment (itsthejoker) [inline tor_ocr/main.py:49]: There's a good chance we'll also need to specify the login API url at runtime as well, not just the base url.

Comment (p-mayor) [reply tor_ocr/main.py:49]: Yep, not sure how that got lost in the refactor, added it back in.

File: tor_ocr/core/blossom.py (hunk 5 above)
Comment (p-mayor) [reply tor_ocr/core/blossom.py]: no problem, should be fixed now

File: tor_ocr/main.py [hunk 7]
```diff
@@ -36,6 +38,15 @@
 
     u_tor_post_id.reply(ocr_magic)
 """
+dotenv.load_dotenv()
+
+b_api = BlossomAPI(
+    email=os.environ.get('TOR_OCR_EMAIL'), 
+    password=os.environ.get('TOR_OCR_PASSWORD'), 
+    api_key='TOR_OCR_BLOSSOM_API_KEY',
```
Comment (NGTmeaty) [inline tor_ocr/main.py]: Is the API key supposed to be obtained from the os environment variables?

Comment (p-mayor) [reply tor_ocr/main.py]: Yes, this was implemented due to direction from @itsthejoker

Comment (itsthejoker) [reply tor_ocr/main.py]: (He's mentioning that that's a bare string, not pulling an environment variable 😉 )

Comment (p-mayor) [reply tor_ocr/main.py]: Woops, nice catch!

Comment (p-mayor) [reply tor_ocr/main.py]: Should be fixed now.

**Comment by TheLonelyGhost** [comment]
@p-mayor I'm not entirely certain if you're still working on this PR or if you've moved onto other stuff in the meantime, but if it's ready to be reviewed again, feel free to hit the ~recycle symbol~ rotating arrows next to a reviewer.

Looks like there may be changes that need to be pulled in from upstream (`master` branch) too.

File: tor_ocr/main.py [hunk 8]
```diff
@@ -260,6 +270,29 @@ def run(config):
 
         thing_to_reply_to = thing_to_reply_to.reply(_(chunk))
 
+        # grab first comment for transcription url
+        if counter==0:
+            first_comment = thing_to_reply_to
+
+    # Get submission from blossom, hard coded for testing purposes
+    blossom_submission = b_api.get("/submission/1/").json()
```
Comment (itsthejoker) [inline tor_ocr/main.py]: Let's remove this and uncomment out the production version :)

File: tor_ocr/main.py [hunk 9]
```diff
@@ -268,12 +301,25 @@ def noop(cfg):
     logging.info('Loop!')
 
 
+def get_volunteer_id(reddit_username):
+    #remove first 2 characters (eg: u/transcribot => transcribot)
+    formatted_name = reddit_username[2:]
```
Comment (itsthejoker) [inline tor_ocr/main.py]: Is it guaranteed that we'll always get a string where the first two characters are `u/`? Let's add a quick check for that; if we also change the arg from `reddit_username` to `username` for length purposes, we can fit a ternary to take care of it:

```suggestion
    formatted_name = username[2:] if username.startswith("u/") else username
```

**Comment by itsthejoker** [comment]
I feel like we're _really close_ with this! Great work, @p-mayor!

File: tor_ocr/main.py (hunk 9 above)
Comment (p-mayor) [reply tor_ocr/main.py]: I like it, updating!

File: tor_ocr/main.py (hunk 8 above)
Comment (p-mayor) [reply tor_ocr/main.py]: Done

File: tor_ocr/core/blossom.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+            params=params
+        )
+
+        for _ in range(3):
+            prepped = self.s.prepare_request(req)
+            settings = self.s.merge_environment_settings(prepped.url, {}, None, None, None)
+            resp = self.s.send(prepped, **settings)
+
+            if resp.status_code == 403:
+                if resp.json().get('detail') == "Authentication credentials were not provided.":
```
Comment (TheLonelyGhost) [inline tor_ocr/core/blossom.py:54]: I like it. I like it a lot. 😄

File: tor_ocr/main.py [hunk 11]
```diff
@@ -245,11 +256,10 @@ def run(config):
         _(base_comment.format(result['process_time_in_ms'] / 1000))
     )
 
-    for chunk in chunks(result['text'], 9000):
+    for counter, chunk in enumerate(chunks(result['text'], 9000)):
```
Comment (TheLonelyGhost) [inline tor_ocr/main.py:259]: > ... `t'], 9000)):`

over9000

**Review by itsthejoker — APPROVED** [review]
Due to changes to blossom, this will need some updates but that is not reflective of the excellent work put in by @p-mayor. Merging into feature branch so that work can continue.

**Pull request merged**